Document & implement the transmutation modeled by `BikeshedIntrinsicFrom`

jswrenn opened the pull request:
Documents that `BikeshedIntrinsicFrom` models transmute-via-union, which is slightly more expressive than the transmute-via-cast implemented by `transmute_copy`. Additionally, we provide an implementation of transmute-via-union as a method on the `BikeshedIntrinsicFrom` trait with additional documentation on the boundary between trait invariants and caller obligations.

Whether or not transmute-via-union is the right kind of transmute to model remains up for discussion [1]. Regardless, it seems wise to document the present behavior.

[1] https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/What.20'kind'.20of.20transmute.20to.20model.3F/near/426331967

Tracking Issue: https://github.com/rust-lang/rust/issues/99571

r? @compiler-errors 

cc @scottmcm, @Lokathor

jswrenn committed:
document & impl the transmutation modeled by `BikeshedIntrinsicFrom`

Documents that `BikeshedIntrinsicFrom` models transmute-via-union,
which is slightly more expressive than the transmute-via-cast
implemented by `transmute_copy`. Additionally, we provide an
implementation of transmute-via-union as a method on the
`BikeshedIntrinsicFrom` trait with additional documentation on
the boundary between trait invariants and caller obligations.

Whether or not transmute-via-union is the right kind of transmute
to model remains up for discussion [1]. Regardless, it seems wise
to document the present behavior.

[1] https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/What.20'kind'.20of.20transmute.20to.20model.3F/near/426331967

the8472 commented on library/core/src/mem/transmutability.rs:
```diff
@@ -1,10 +1,63 @@
+///
+/// ## Portability
+///
+/// Implementations of this trait do provide any guarantee of portability across
```
```suggestion
/// Implementations of this trait do not provide any guarantee of portability across
```

typo?

jswrenn replied:
[Fixed!](https://github.com/rust-lang/rust/compare/b806b5e052cd78ffedcffb6484c62cde9250ac7c..96cfefbb125916871760c1df4c2ee29944635e90) Thanks!

joshlf commented on library/core/src/mem/transmutability.rs:
```diff
@@ -13,6 +66,53 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
+    /// # Safety
+    ///
+    /// The safety obligations of the caller depend on the value of `ASSUME`:
+    /// - If [`ASSUME.alignment`](Assume::alignment), the caller must prove that
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have well-aligned referents upon transmutation into
+    ///   `Self`.
```
IIUC, this is assuming that the only types that may be transmuted into references are other references, and thus the set of references in `src` is a (possibly trivial) super-set of the set of references in the resulting `Self`. It might be more straightforward and more future-proof to not make this assumption and only refer directly to the set of references in `Self`.

Also, IMO the wording "well-aligned referent" is confusing since it makes it seem like it's a property of the content of the referent rather than a property of the *address* of the referent. There are different ways you could rephrase this, but I'd suggest using a phrasing that treats alignment as a property of the reference, not a property of the referent.

jswrenn replied:
Let's make further comments about this here, to keep things centralized: https://github.com/rust-lang/rust/pull/129032#discussion_r1714491421

jswrenn replied:
I worry that "well-aligned reference" suggests that the address of the reference is well-aligned, rather than the address of the referent. What do you think of:
```suggestion
    /// - If [`ASSUME.alignment`](Assume::alignment), the caller must guarantee
    ///   that references in the returned `Self` have well-aligned referents.
```

joshlf replied:
Maybe to be slightly more verbose:

> If `ASSUME.alignment`, the caller must guarantee that the addresses of references in the returned `Self` satisfy the alignment requirements of their referent types.

Or alternatively:

> If `ASSUME.alignment`, the caller must guarantee that the addresses of references in the returned `Self` are properly aligned for their referent types.

(The phrase "properly aligned" in my mind implies a correctness property which is binary - it's either satisfied or not. By contrast, "well-aligned" feels like it's describing something on a scale, or something that is subjective. That's a tiny nit and not important.)

jswrenn replied:
[Done](https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR90-R92)

joshlf commented on library/core/src/mem/transmutability.rs:
```diff
@@ -13,6 +66,53 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have well-aligned referents upon transmutation into
+    ///   `Self`.
+    /// - If [`ASSUME.lifetimes`](Assume::lifetimes), the caller must prove that
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have referents that live as long as `Self`.
```
Same comment here about referring to objects in `src` vs in the resulting `Self`. However, this gets trickier since the rules for what constitute valid lifetimes are super tricky. My first inclination was to punt and say something that just passes the buck to Rust, e.g.:

> the caller must prove that it is sound to transmute `src` into `Self` with the given lifetimes

...but of course, that demands that the caller prove the soundness of the *entire* transmutation, not just the lifetime aspects of the transmutation. I'm honestly not sure how you'd express the idea of "the lifetime aspects of a transmutation" in precise language.

At a minimum, any such language would need to refer to the referents of references since references are valid for particular lifetimes *in virtue of* what happens with their referents in other parts of the program. I'm thinking of something similar to the language we use regarding reference validity in zerocopy's `Ptr` type, where we talk a lot about the referent.

jswrenn replied:
Hm, yeah, this needs substantial reworking. The intention of this `Assume` switch is to permit using this trait/method for lifetime extension. A common case of this is extending lifetimes to `'static`, even though such lifetimes rarely accurately describe the referent.

I think your inclination to phrase this obligation with respect to what happens *after* the transmutation is correct. (And a similar complaint can be levied against `Assume::safety`.)

jswrenn replied:
@joshlf, what do you think of:
```suggestion
    /// - If [`ASSUME.lifetimes`](Assume::lifetimes), the caller must guarantee
    ///   that pointers in the returned `Self` will not dangle.
```
This is intentionally conservative, allowing for the language validity rules of references to be relaxed in the future. For instance, it may eventually be decided that non-dangling-ness is a library invariant, rather than a language invariant. If that happens, we can relax this obligation to "the caller must guarantee that no dangling dereferences occur through the returned `Self`" without breaking any existing proofs.

joshlf replied:
I like that, although two tiny nits:
- Instead of "pointers", I would say "references"
- Instead of "will not dangle", I would say "will not outlive their referents"

jswrenn replied:
[Done](https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR90-R92)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10361029773/job/28680682875) [(plain)](https://github.com/rust-lang/rust/commit/96cfefbb125916871760c1df4c2ee29944635e90/checks/28680682875/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:3aacb9c90579defe09351ac5e8ee504359f8054da6326ff19038f1b7c90e3cb2aafe33685c6d9b76ee8d2ccbd187ca80c46ab5380485abdd8c0ce7d69cd8d8fd:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/closures/coerce-unsafe-to-closure.rs stdout ----
diff of stderr:

- error[E0277]: expected a `FnOnce(&str)` closure, found `unsafe extern "rust-intrinsic" fn(_) -> _ {transmute::<_, _>}`
+ error[E0277]: expected a `FnOnce(&str)` closure, found `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}`
3    |
3    |
4 LL |     let x: Option<&[u8]> = Some("foo").map(std::mem::transmute);
6    |                                        |
7    |                                        required by a bound introduced by this call
8    |
8    |
-    = help: the trait `FnOnce(&str)` is not implemented for fn item `unsafe extern "rust-intrinsic" fn(_) -> _ {transmute::<_, _>}`
+    = help: the trait `FnOnce(&str)` is not implemented for fn item `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}`
10    = note: unsafe function cannot be called generically without an unsafe block
11 note: required by a bound in `Option::<T>::map`
12   --> $SRC_DIR/core/src/option.rs:LL:COL

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/coerce-unsafe-to-closure/coerce-unsafe-to-closure.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args closures/coerce-unsafe-to-closure.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/closures/coerce-unsafe-to-closure.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/coerce-unsafe-to-closure" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/coerce-unsafe-to-closure/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0277]: expected a `FnOnce(&str)` closure, found `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}`
   |
   |
LL |     let x: Option<&[u8]> = Some("foo").map(std::mem::transmute);
   |                                        --- ^^^^^^^^^^^^^^^^^^^ call the function in a closure: `|| unsafe { /* code */ }`
   |                                        required by a bound introduced by this call
   |
   |
   = help: the trait `FnOnce(&str)` is not implemented for fn item `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}`
   = note: unsafe function cannot be called generically without an unsafe block
note: required by a bound in `Option::<T>::map`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
---
diff of stderr:

7    |            expected due to this
8    |
9    = note: expected fn pointer `unsafe extern "rust-intrinsic" fn(isize) -> usize`
-                  found fn item `unsafe extern "rust-intrinsic" fn(_) -> _ {transmute::<_, _>}`
+                  found fn item `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}`
11 
- error[E0606]: casting `unsafe extern "rust-intrinsic" fn(_) -> _ {transmute::<_, _>}` as `unsafe extern "rust-intrinsic" fn(isize) -> usize` is invalid
+ error[E0606]: casting `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}` as `unsafe extern "rust-intrinsic" fn(isize) -> usize` is invalid
14    |
14    |
15 LL |     let _ = std::mem::transmute as unsafe extern "rust-intrinsic" fn(isize) -> usize;

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/intrinsics/reify-intrinsic/reify-intrinsic.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args intrinsics/reify-intrinsic.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/intrinsics/reify-intrinsic.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/intrinsics/reify-intrinsic" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/intrinsics/reify-intrinsic/auxiliary"
--- stderr -------------------------------
error[E0308]: cannot coerce intrinsics to function pointers
##[error]  --> /checkout/tests/ui/intrinsics/reify-intrinsic.rs:6:64
   |
   |
LL |     let _: unsafe extern "rust-intrinsic" fn(isize) -> usize = std::mem::transmute;
   |            -------------------------------------------------   ^^^^^^^^^^^^^^^^^^^ cannot coerce intrinsics to function pointers
   |            expected due to this
   |
   |
   = note: expected fn pointer `unsafe extern "rust-intrinsic" fn(isize) -> usize`
                 found fn item `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}`

error[E0606]: casting `unsafe extern "rust-intrinsic" fn(_) -> _ {std::intrinsics::transmute::<_, _>}` as `unsafe extern "rust-intrinsic" fn(isize) -> usize` is invalid
   |
   |
LL |     let _ = std::mem::transmute as unsafe extern "rust-intrinsic" fn(isize) -> usize;

error[E0308]: cannot coerce intrinsics to function pointers
##[error]  --> /checkout/tests/ui/intrinsics/reify-intrinsic.rs:17:9
   |
   |
LL |         std::intrinsics::likely, //~ ERROR cannot coerce
   |         ^^^^^^^^^^^^^^^^^^^^^^^ cannot coerce intrinsics to function pointers
   |
   = note: expected fn pointer `unsafe extern "rust-intrinsic" fn(_) -> _`
                 found fn item `fn(_) -> _ {likely}`
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0308, E0606.
For more information about an error, try `rustc --explain E0308`.
```

</details>

compiler-errors commented on library/core/src/mem/transmutability.rs:
```diff
@@ -1,10 +1,63 @@
+/// This trait cannot be implemented explicitly. It is implemented on-the-fly by
+/// the compiler for all types `Src` and `Self` such that, given a set of safety
+/// obligations on the programmer (see [`Assume`]), the compiler has proved that
+/// the bits of a value of type `Src` can be soundly reinterpreted as `Self`.
```
I don't actually agree with this change. "as `Self`" sounds more correct in my idiolect.

joshlf replied:
My thinking is that the left-hand side refers to a runtime object - "a value of type `Src`" - and so the right-hand side should too. Conservatively, we could say "a value of type `Self`", but IIUC "a `Self`" is a common shorthand for that.

Lokathor replied:
If we take the `Self` part out for a moment, it would be common to say "as an i32", or "as an array", so "a `Self`" probably fits.

jswrenn replied:
I'm inclined to agree with @Lokathor's argument, so I've [opted](https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR10) for "a `Self`".

jswrenn commented on library/core/src/mem/transmutability.rs:
```diff
@@ -13,6 +66,53 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
+    /// When satisfying the above obligations (if any), the caller must *not*
+    /// assume that this trait provides any inherent guarantee of layout
+    /// [portability](#portability) or [stability](#stability).
+    unsafe fn transmute(src: Src) -> Self
```
@compiler-errors, I'm getting an ICE when I try to call this method in tests:
```
---- core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41) stdout ----
error: internal compiler error: compiler/rustc_middle/src/ty/instance.rs:599:21: failed to resolve instance for <Dst as BikeshedIntrinsicFrom<u8, Assume { alignment: false, lifetimes: false, safety: true, validity: false }>>::transmute
  --> core/src/mem/transmutability.rs:51:5
   |
13 |     <Dst as BikeshedIntrinsicFrom<u8, { Assume::SAFETY }>>::transmute(src)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at compiler/rustc_middle/src/ty/instance.rs:599:21:
Box<dyn Any>
stack backtrace:
   0:     0x7f11424685b6 - trace
                               at /home/ubuntu/projects/rust/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x7f11424685b6 - trace_unsynchronized<std::sys::backtrace::_print_fmt::{closure_env#1}>
                               at /home/ubuntu/projects/rust/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f11424685b6 - _print_fmt
                               at /home/ubuntu/projects/rust/library/std/src/sys/backtrace.rs:66:9
   3:     0x7f11424685b6 - fmt
                               at /home/ubuntu/projects/rust/library/std/src/sys/backtrace.rs:39:26
   4:     0x7f11424c82eb - fmt
                               at /home/ubuntu/projects/rust/library/core/src/fmt/rt.rs:173:76
[...]
```
https://github.com/rust-lang/rust/blob/80eb5a8e910e5185d47cdefe3732d839c78a5e7e/compiler/rustc_middle/src/ty/instance.rs#L599-L603

Do you know if there any special ceremony needed to add a defaulted method to a compiler-implemented trait?

compiler-errors replied:
Add a case to the `traits::ImplSource::Builtin(BuiltinImplSource::Misc, _)` arm in `resolve_associated_item` in `compiler/rustc_ty_utils/src/instance.rs`. Look for the `clone_from` case, for example.

jswrenn replied:
Thanks! [Done!](https://github.com/rust-lang/rust/pull/129032/files#diff-a54c982c1e255569ddede4404cdf85ecfbd93c20edfa8a090a5f2dc2bc1dff6b)

jswrenn commented on library/core/src/mem/transmutability.rs:
```diff
@@ -1,10 +1,63 @@
+///     union Transmute<T, U> {
+///         src: ManuallyDrop<T>,
+///         dst: ManuallyDrop<U>,
+///     }
```
I hadn't put that much thought into this, actually. I don't think there's any risk of confusion from shadowing `Src` and `Dst` (even if you didn't realize they were shadowed, the meaning of the code wouldn't change), so I've opted to just use `Src` and `Dst`:
- https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR21-R23
- https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR114-R116

jswrenn commented on library/core/src/mem/transmutability.rs:
```diff
@@ -1,10 +1,63 @@
+/// trait may also be non-deterministic. Even if `Src` and `Self` have
+/// deterministic layouts (e.g., they are `repr(C)` structs), Rust does not
+/// specify the alignments of its primitive integer types, and layouts that
+/// involve these types may vary across toolchains.
```
Good point! This now says "toolchains, targets or compilations".

https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR59

joshlf replied:
One change left:

> Rust does not specify the alignments of its primitive integer types, and layouts that involve these types may vary across toolchains.

This should say "across toolchains or targets" or something similar.

jswrenn replied:
[Done!](https://github.com/rust-lang/rust/blob/ac7b5526dedbab6ea511c3566292ea2bebdb8931/library/core/src/mem/transmutability.rs#L64-L66)

jswrenn commented on library/core/src/mem/transmutability.rs:
```diff
@@ -1,10 +1,63 @@
+///
+/// Note that this construction supports some transmutations forbidden by
+/// [`mem::transmute_copy`](super::transmute_copy), namely transmutations that
+/// extend the trailing padding of `Src` to fill `Dst`.
```
I've reworded this slightly, and included an example provided by @Lokathor on Zulip: https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR59

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10389265611/job/28766900753) [(plain)](https://github.com/rust-lang/rust/commit/1c0e7fac16490ea2e83ab0127c364231a42742ad/checks/28766900753/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.898 Building wheels for collected packages: reuse
#16 2.899   Building wheel for reuse (pyproject.toml): started
#16 3.152   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.153   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.153   Stored in directory: /tmp/pip-ephem-wheel-cache-eyrxt9s5/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.155 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.559 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.559 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.7s
---
........................................................................................  704/4880
................................................................................iiii....  792/4880
.................................iiii...................................................  880/4880
..........................................
core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41) ... F
........................................................................................ 1056/4880
........................................................................................ 1144/4880
........................................................................................ 1232/4880
........................................................................................ 1320/4880
---
...................................i....

failures:

---- core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41) stdout ----
##[error]error: internal compiler error: compiler/rustc_middle/src/ty/instance.rs:598:21: failed to resolve instance for <Dst as BikeshedIntrinsicFrom<u8, Assume { alignment: false, lifetimes: false, safety: true, validity: false }>>::transmute
  --> core/src/mem/transmutability.rs:51:5
   |
13 |     <Dst as BikeshedIntrinsicFrom<u8, { Assume::SAFETY }>>::transmute(src)

thread 'rustc' panicked at compiler/rustc_middle/src/ty/instance.rs:598:21:
Box<dyn Any>
stack backtrace:
---
   8:     0x7fc1ae78eb91 - std::panicking::default_hook::{{closure}}::hf6a2255341957303
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:269:22
   9:     0x7fc1ae78e86c - std::panicking::default_hook::h0b7d4412c1e361bd
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:296:9
  10:     0x7fc1b1a9ea50 - std[efeefe75b6db041f]::panicking::update_hook::<alloc[759981cd4e4e801b]::boxed::Box<rustc_driver_impl[c3312e73a30af321]::install_ice_hook::{closure#0}>>::{closure#0}
  11:     0x7fc1ae78f55f - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hdc9e661750688cd9
  12:     0x7fc1ae78f55f - std::panicking::rust_panic_with_hook::h6fa5f6f38389e656
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:808:13
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:808:13
  13:     0x7fc1b1ad6121 - std[efeefe75b6db041f]::panicking::begin_panic::<rustc_errors[dec2fc3ea9d24357]::ExplicitBug>::{closure#0}
  14:     0x7fc1b1ac99e6 - std[efeefe75b6db041f]::sys::backtrace::__rust_end_short_backtrace::<std[efeefe75b6db041f]::panicking::begin_panic<rustc_errors[dec2fc3ea9d24357]::ExplicitBug>::{closure#0}, !>
  15:     0x7fc1b1ac9996 - std[efeefe75b6db041f]::panicking::begin_panic::<rustc_errors[dec2fc3ea9d24357]::ExplicitBug>
  16:     0x7fc1b1adf7b1 - <rustc_errors[dec2fc3ea9d24357]::diagnostic::BugAbort as rustc_errors[dec2fc3ea9d24357]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  17:     0x7fc1b2023b7d - <rustc_errors[dec2fc3ea9d24357]::DiagCtxtHandle>::span_bug::<rustc_span[e52881666e69bc1d]::span_encoding::Span, alloc[759981cd4e4e801b]::string::String>
  18:     0x7fc1b20c5628 - rustc_middle[2cecd0bcfd982753]::util::bug::opt_span_bug_fmt::<rustc_span[e52881666e69bc1d]::span_encoding::Span>::{closure#0}
  19:     0x7fc1b20ab18a - rustc_middle[2cecd0bcfd982753]::ty::context::tls::with_opt::<rustc_middle[2cecd0bcfd982753]::util::bug::opt_span_bug_fmt<rustc_span[e52881666e69bc1d]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  20:     0x7fc1b20ab02b - rustc_middle[2cecd0bcfd982753]::ty::context::tls::with_context_opt::<rustc_middle[2cecd0bcfd982753]::ty::context::tls::with_opt<rustc_middle[2cecd0bcfd982753]::util::bug::opt_span_bug_fmt<rustc_span[e52881666e69bc1d]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  21:     0x7fc1b20c5557 - rustc_middle[2cecd0bcfd982753]::util::bug::span_bug_fmt::<rustc_span[e52881666e69bc1d]::span_encoding::Span>
  22:     0x7fc1b343b333 - <rustc_middle[2cecd0bcfd982753]::ty::instance::Instance>::expect_resolve
  23:     0x7fc1b3b1efc2 - rustc_monomorphize[cddd32943e0cc721]::collector::collect_items_rec::{closure#0}
  24:     0x7fc1b3b1bdf7 - rustc_monomorphize[cddd32943e0cc721]::collector::collect_items_rec
  25:     0x7fc1b3b1c51c - rustc_monomorphize[cddd32943e0cc721]::collector::collect_items_rec
  26:     0x7fc1b34c547a - rustc_monomorphize[cddd32943e0cc721]::partitioning::collect_and_partition_mono_items
  27:     0x7fc1b409f724 - rustc_query_impl[31fd7ff07bfdf275]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[31fd7ff07bfdf275]::query_impl::collect_and_partition_mono_items::dynamic_query::{closure#2}::{closure#0}, rustc_middle[2cecd0bcfd982753]::query::erase::Erased<[u8; 24usize]>>
  28:     0x7fc1b409f709 - <rustc_query_impl[31fd7ff07bfdf275]::query_impl::collect_and_partition_mono_items::dynamic_query::{closure#2} as core[c78a77099d0262f3]::ops::function::FnOnce<(rustc_middle[2cecd0bcfd982753]::ty::context::TyCtxt, ())>>::call_once
  29:     0x7fc1b409f391 - rustc_query_system[e7fb6eb8ab67ecdb]::query::plumbing::try_execute_query::<rustc_query_impl[31fd7ff07bfdf275]::DynamicConfig<rustc_query_system[e7fb6eb8ab67ecdb]::query::caches::SingleCache<rustc_middle[2cecd0bcfd982753]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[31fd7ff07bfdf275]::plumbing::QueryCtxt, false>
  30:     0x7fc1b409f0f1 - rustc_query_impl[31fd7ff07bfdf275]::query_impl::collect_and_partition_mono_items::get_query_non_incr::__rust_end_short_backtrace
  31:     0x7fc1b3f12176 - <rustc_codegen_llvm[49d4f4c542a8fa73]::LlvmCodegenBackend as rustc_codegen_ssa[e11e52d96a357390]::traits::backend::CodegenBackend>::codegen_crate
  32:     0x7fc1b3fbf25f - <rustc_interface[b6630b0d0d781590]::queries::Linker>::codegen_and_build_linker
  33:     0x7fc1b3d24142 - rustc_interface[b6630b0d0d781590]::interface::run_compiler::<core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>, rustc_driver_impl[c3312e73a30af321]::run_compiler::{closure#0}>::{closure#1}
  34:     0x7fc1b3c6351b - std[efeefe75b6db041f]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b6630b0d0d781590]::util::run_in_thread_with_globals<rustc_interface[b6630b0d0d781590]::interface::run_compiler<core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>, rustc_driver_impl[c3312e73a30af321]::run_compiler::{closure#0}>::{closure#1}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>
  35:     0x7fc1b3c632ea - <<std[efeefe75b6db041f]::thread::Builder>::spawn_unchecked_<rustc_interface[b6630b0d0d781590]::util::run_in_thread_with_globals<rustc_interface[b6630b0d0d781590]::interface::run_compiler<core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>, rustc_driver_impl[c3312e73a30af321]::run_compiler::{closure#0}>::{closure#1}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>::{closure#1} as core[c78a77099d0262f3]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7fc1ae7995bb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5e49ad344eded169
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/alloc/src/boxed.rs:2070:9
  37:     0x7fc1ae7995bb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc344207bfeabc120
  38:     0x7fc1ae7995bb - std::sys::pal::unix::thread::Thread::new::thread_start::h30a261106ad4495b
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/sys/pal/unix/thread.rs:108:17
  39:     0x7fc1ae53dac3 - <unknown>
  40:     0x7fc1ae5cf850 - <unknown>
  40:     0x7fc1ae5cf850 - <unknown>
  41:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

warning: the ICE couldn't be written to `/checkout/library/rustc-ice-2024-08-14T14_33_03-60930.txt`: Read-only file system (os error 30)
note: rustc 1.81.0-beta.2 (08328a323 2024-07-25) running on x86_64-unknown-linux-gnu

query stack during panic:
query stack during panic:
#0 [collect_and_partition_mono_items] collect_and_partition_mono_items
error: aborting due to 1 previous error

Couldn't compile the test.


failures:
    core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41)
test result: FAILED. 4835 passed; 1 failed; 44 ignored; 0 measured; 0 filtered out; finished in 185.07s

error: doctest failed, to rerun pass `-p core --doc`
Build completed unsuccessfully in 0:05:18
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10390909435/job/28772522202) [(plain)](https://github.com/rust-lang/rust/commit/ac7b5526dedbab6ea511c3566292ea2bebdb8931/checks/28772522202/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.639 Building wheels for collected packages: reuse
#16 2.640   Building wheel for reuse (pyproject.toml): started
#16 2.882   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 2.883   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 2.884   Stored in directory: /tmp/pip-ephem-wheel-cache-cie80l78/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 2.887 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.278 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.279 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.4s
---
........................................................................................  704/4880
................................................................................iiii....  792/4880
.................................iiii...................................................  880/4880
.........................................
core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41) ... F
........................................................................................ 1056/4880
........................................................................................ 1144/4880
........................................................................................ 1232/4880
........................................................................................ 1320/4880
---
...................................i....

failures:

---- core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41) stdout ----
##[error]error: internal compiler error: compiler/rustc_middle/src/ty/instance.rs:598:21: failed to resolve instance for <Dst as BikeshedIntrinsicFrom<u8, Assume { alignment: false, lifetimes: false, safety: true, validity: false }>>::transmute
  --> core/src/mem/transmutability.rs:50:18
   |
12 |   let _ = unsafe { <Dst as BikeshedIntrinsicFrom<u8, { Assume::SAFETY
   |  __________________^
13 | |     }>>::transmute(src) };

thread 'rustc' panicked at compiler/rustc_middle/src/ty/instance.rs:598:21:
Box<dyn Any>
stack backtrace:
---
   8:     0x7f8491f8eb91 - std::panicking::default_hook::{{closure}}::hf6a2255341957303
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:269:22
   9:     0x7f8491f8e86c - std::panicking::default_hook::h0b7d4412c1e361bd
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:296:9
  10:     0x7f849529ea50 - std[efeefe75b6db041f]::panicking::update_hook::<alloc[759981cd4e4e801b]::boxed::Box<rustc_driver_impl[c3312e73a30af321]::install_ice_hook::{closure#0}>>::{closure#0}
  11:     0x7f8491f8f55f - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hdc9e661750688cd9
  12:     0x7f8491f8f55f - std::panicking::rust_panic_with_hook::h6fa5f6f38389e656
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:808:13
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/panicking.rs:808:13
  13:     0x7f84952d6121 - std[efeefe75b6db041f]::panicking::begin_panic::<rustc_errors[dec2fc3ea9d24357]::ExplicitBug>::{closure#0}
  14:     0x7f84952c99e6 - std[efeefe75b6db041f]::sys::backtrace::__rust_end_short_backtrace::<std[efeefe75b6db041f]::panicking::begin_panic<rustc_errors[dec2fc3ea9d24357]::ExplicitBug>::{closure#0}, !>
  15:     0x7f84952c9996 - std[efeefe75b6db041f]::panicking::begin_panic::<rustc_errors[dec2fc3ea9d24357]::ExplicitBug>
  16:     0x7f84952df7b1 - <rustc_errors[dec2fc3ea9d24357]::diagnostic::BugAbort as rustc_errors[dec2fc3ea9d24357]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  17:     0x7f8495823b7d - <rustc_errors[dec2fc3ea9d24357]::DiagCtxtHandle>::span_bug::<rustc_span[e52881666e69bc1d]::span_encoding::Span, alloc[759981cd4e4e801b]::string::String>
  18:     0x7f84958c5628 - rustc_middle[2cecd0bcfd982753]::util::bug::opt_span_bug_fmt::<rustc_span[e52881666e69bc1d]::span_encoding::Span>::{closure#0}
  19:     0x7f84958ab18a - rustc_middle[2cecd0bcfd982753]::ty::context::tls::with_opt::<rustc_middle[2cecd0bcfd982753]::util::bug::opt_span_bug_fmt<rustc_span[e52881666e69bc1d]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  20:     0x7f84958ab02b - rustc_middle[2cecd0bcfd982753]::ty::context::tls::with_context_opt::<rustc_middle[2cecd0bcfd982753]::ty::context::tls::with_opt<rustc_middle[2cecd0bcfd982753]::util::bug::opt_span_bug_fmt<rustc_span[e52881666e69bc1d]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  21:     0x7f84958c5557 - rustc_middle[2cecd0bcfd982753]::util::bug::span_bug_fmt::<rustc_span[e52881666e69bc1d]::span_encoding::Span>
  22:     0x7f8496c3b333 - <rustc_middle[2cecd0bcfd982753]::ty::instance::Instance>::expect_resolve
  23:     0x7f849731efc2 - rustc_monomorphize[cddd32943e0cc721]::collector::collect_items_rec::{closure#0}
  24:     0x7f849731bdf7 - rustc_monomorphize[cddd32943e0cc721]::collector::collect_items_rec
  25:     0x7f849731c51c - rustc_monomorphize[cddd32943e0cc721]::collector::collect_items_rec
  26:     0x7f8496cc547a - rustc_monomorphize[cddd32943e0cc721]::partitioning::collect_and_partition_mono_items
  27:     0x7f849789f724 - rustc_query_impl[31fd7ff07bfdf275]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[31fd7ff07bfdf275]::query_impl::collect_and_partition_mono_items::dynamic_query::{closure#2}::{closure#0}, rustc_middle[2cecd0bcfd982753]::query::erase::Erased<[u8; 24usize]>>
  28:     0x7f849789f709 - <rustc_query_impl[31fd7ff07bfdf275]::query_impl::collect_and_partition_mono_items::dynamic_query::{closure#2} as core[c78a77099d0262f3]::ops::function::FnOnce<(rustc_middle[2cecd0bcfd982753]::ty::context::TyCtxt, ())>>::call_once
  29:     0x7f849789f391 - rustc_query_system[e7fb6eb8ab67ecdb]::query::plumbing::try_execute_query::<rustc_query_impl[31fd7ff07bfdf275]::DynamicConfig<rustc_query_system[e7fb6eb8ab67ecdb]::query::caches::SingleCache<rustc_middle[2cecd0bcfd982753]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[31fd7ff07bfdf275]::plumbing::QueryCtxt, false>
  30:     0x7f849789f0f1 - rustc_query_impl[31fd7ff07bfdf275]::query_impl::collect_and_partition_mono_items::get_query_non_incr::__rust_end_short_backtrace
  31:     0x7f8497712176 - <rustc_codegen_llvm[49d4f4c542a8fa73]::LlvmCodegenBackend as rustc_codegen_ssa[e11e52d96a357390]::traits::backend::CodegenBackend>::codegen_crate
  32:     0x7f84977bf25f - <rustc_interface[b6630b0d0d781590]::queries::Linker>::codegen_and_build_linker
  33:     0x7f8497524142 - rustc_interface[b6630b0d0d781590]::interface::run_compiler::<core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>, rustc_driver_impl[c3312e73a30af321]::run_compiler::{closure#0}>::{closure#1}
  34:     0x7f849746351b - std[efeefe75b6db041f]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b6630b0d0d781590]::util::run_in_thread_with_globals<rustc_interface[b6630b0d0d781590]::interface::run_compiler<core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>, rustc_driver_impl[c3312e73a30af321]::run_compiler::{closure#0}>::{closure#1}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>
  35:     0x7f84974632ea - <<std[efeefe75b6db041f]::thread::Builder>::spawn_unchecked_<rustc_interface[b6630b0d0d781590]::util::run_in_thread_with_globals<rustc_interface[b6630b0d0d781590]::interface::run_compiler<core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>, rustc_driver_impl[c3312e73a30af321]::run_compiler::{closure#0}>::{closure#1}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c78a77099d0262f3]::result::Result<(), rustc_span[e52881666e69bc1d]::ErrorGuaranteed>>::{closure#1} as core[c78a77099d0262f3]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7f8491f995bb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5e49ad344eded169
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/alloc/src/boxed.rs:2070:9
  37:     0x7f8491f995bb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc344207bfeabc120
  38:     0x7f8491f995bb - std::sys::pal::unix::thread::Thread::new::thread_start::h30a261106ad4495b
                               at /rustc/08328a323ecd80b443a8fcc72c6b1071f48e233f/library/std/src/sys/pal/unix/thread.rs:108:17
  39:     0x7f8491d3dac3 - <unknown>
  40:     0x7f8491dcf850 - <unknown>
  40:     0x7f8491dcf850 - <unknown>
  41:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

warning: the ICE couldn't be written to `/checkout/library/rustc-ice-2024-08-14T16_14_51-60930.txt`: Read-only file system (os error 30)
note: rustc 1.81.0-beta.2 (08328a323 2024-07-25) running on x86_64-unknown-linux-gnu

query stack during panic:
query stack during panic:
#0 [collect_and_partition_mono_items] collect_and_partition_mono_items
error: aborting due to 1 previous error

Couldn't compile the test.


failures:
    core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41)
test result: FAILED. 4835 passed; 1 failed; 44 ignored; 0 measured; 0 filtered out; finished in 184.41s

error: doctest failed, to rerun pass `-p core --doc`
Build completed unsuccessfully in 0:05:19
```

</details>

joshlf commented on library/core/src/mem/transmutability.rs:
```diff
@@ -13,28 +83,96 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
-    /// When `true`, the compiler assumes that *you* have ensured that no
-    /// unsoundness will arise from violating the safety invariants of the
-    /// destination type (and sometimes of the source type, too).
+    /// When `false`,  [`BikeshedIntrinsicFrom`] is not implemented for
```
```suggestion
    /// When `false`, [`BikeshedIntrinsicFrom`] is not implemented for
```

jswrenn replied:
[Done](https://github.com/rust-lang/rust/compare/8bf3df91b8be5d0a88bda83715a6db795d346058..7254d9d761afb649b646f16a469683c4ebc47f9a#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfL161)

joshlf commented on library/core/src/mem/transmutability.rs:
```diff
@@ -13,28 +83,96 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
 
-    /// When `true`, the compiler assumes that *you* are ensuring that the source type is actually a valid
-    /// instance of the destination type.
+    /// When `false`, [`BikeshedIntrinsicFrom`] is not implemented for
+    /// transmutations that might violate the language-level bit-validity
+    /// invariant of the return type.
+    ///
+    /// When `true`, [`BikeshedIntrinsicFrom`] assumes that *you* have ensured
+    /// that the value being transmuted is a bit-valid instance of the return
+    /// type.
```
Same here regarding "return"

joshlf commented on library/core/src/mem/transmutability.rs:
```diff
@@ -13,28 +83,96 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
-    /// When `true`, the compiler assumes that *you* have ensured that no
-    /// unsoundness will arise from violating the safety invariants of the
-    /// destination type (and sometimes of the source type, too).
+    /// When `false`,  [`BikeshedIntrinsicFrom`] is not implemented for
+    /// transmutations that might violate the library safety invariants of the
+    /// return type.
+    ///
+    /// When `true`, [`BikeshedIntrinsicFrom`] assumes that *you* have ensured
+    /// that undefined behavior does not arise from using the returned value.
```
Maybe "destination type" or something? "Return" implies we're talking in the context of the `transmute` method rather than in the context of using `BikeshedIntrinsicFrom` as a trait bound.

Same for "returned value" in the second sentence.

jswrenn replied:
[I now use "destination type" and "transmuted value".](https://github.com/rust-lang/rust/compare/8bf3df91b8be5d0a88bda83715a6db795d346058..7254d9d761afb649b646f16a469683c4ebc47f9a#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR161-R166)

joshlf reviewed (approved):
A few optional changes requested, but LGTM!

jswrenn commented on library/core/src/mem/transmutability.rs:
````diff
@@ -1,10 +1,80 @@
+/// }
+/// ```
+///
+/// Note that this construction supports some transmutations forbidden by
+/// [`mem::transmute_copy`](super::transmute_copy); namely, transmutations that
+/// extend the bits of `Src` with trailing padding to fill trailing
+/// uninitialized bytes of `Self`; e.g.:
````
Warning about `mem::transmute_copy` is sufficient, since `mem::transmute` guards against this at compile-time — it's not forbidden in the sense that it's a safety precondition.

joshlf replied:
I see. In that case, maybe clarify that you are referring specifically to the safety precondition as opposed to just generally speaking about what is supported? That distinction wasn’t clear to me from the prose as written.

jswrenn replied:
[Done](https://github.com/rust-lang/rust/compare/8bf3df91b8be5d0a88bda83715a6db795d346058..7254d9d761afb649b646f16a469683c4ebc47f9a#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR36)

joshlf commented:
I see. In that case, maybe clarify that you are referring specifically to
the safety precondition as opposed to just generally speaking about what is
supported? That distinction wasn’t clear to me from the prose as written.

On Fri, Aug 16, 2024 at 10:51 AM Jack Wrenn ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In library/core/src/mem/transmutability.rs
> <https://github.com/rust-lang/rust/pull/129032#discussion_r1720141651>:
>
> > +/// Note that this construction supports some transmutations forbidden by
> +/// [`mem::transmute_copy`](super::transmute_copy); namely, transmutations that
> +/// extend the bits of `Src` with trailing padding to fill trailing
> +/// uninitialized bytes of `Self`; e.g.:
>
> Warning about mem::transmute_copy is sufficient, since mem::transmute
> guards against this at compile-time — it's not forbidden in the sense that
> it's a safety precondition.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/129032#discussion_r1720141651>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH7ML4YWTYPGONWWHR6EW3ZRY33HAVCNFSM6AAAAABMNDHD6OVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDENBTGIZDMNBRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

compiler-errors commented on library/core/src/mem/transmutability.rs:
````diff
@@ -1,10 +1,80 @@
+/// that extend the bits of `Src` with trailing padding to fill trailing
+/// uninitialized bytes of `Self`; e.g.:
+///
+/// ```ignore (cannot doctest this until bootstrap; remove ignore and corresponding test in tests/mem.rs in the future)
````
I believe you can use doc comments' desugared form to gate this correctly.

```suggestion
#[cfg_attr(bootstrap), doc = "```rust,ignore")]
#[cfg_attr(not(bootstrap)), doc = "```rust")]
```

jswrenn replied:
Neat! [Done!](https://github.com/rust-lang/rust/compare/7254d9d761afb649b646f16a469683c4ebc47f9a..60cfa88d3f6f37ae0ce2ebd4ac4ece3aa9538a27)

bors commented:
:umbrella: The latest upstream changes (presumably #122551) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jswrenn commented:
@joshlf, could you take another look at this?
- [I've added a `# Safety` section `BikeshedIntrinsicFrom`](https://github.com/rust-lang/rust/compare/c3ebec939e0dc84e3f230c6f96edd7e54cc9cd42..5ab5080f5a8f481fce864aab470a3bdca8ab2e0c#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR12)
- [I've added doctests to *most* of the `Assume` flags.](https://github.com/rust-lang/rust/compare/c3ebec939e0dc84e3f230c6f96edd7e54cc9cd42..5ab5080f5a8f481fce864aab470a3bdca8ab2e0c#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR151)
  - No examples yet for `Assume::lifetimes`, which is both hard to give a succinct example of, and also so subtly dangerous I hesitate to give an example of it.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10514812106/job/29133556840) [(plain)](https://github.com/rust-lang/rust/commit/5ab5080f5a8f481fce864aab470a3bdca8ab2e0c/checks/29133556840/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.64s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- core/src/mem/transmutability.rs - mem::transmutability::Assume::safety (line 229) stdout ----
error: unexpected end of macro invocation
##[error]  --> core/src/mem/transmutability.rs:249:56
   |
23 | assert_eq!(matches!(maybe_dst, Some(EvenU8 { val: 2 })));
   |                                                        ^ missing tokens in macro arguments
note: while trying to match `,`
  --> /checkout/library/core/src/macros/mod.rs:43:16
   |
   |
43 |     ($left:expr, $right:expr $(,)?) => {

error: aborting due to 1 previous error

Couldn't compile the test.
```

</details>

joshlf commented on library/core/src/mem/transmutability.rs:
```diff
@@ -1,10 +1,88 @@
+///
+/// # Safety
+///
+/// If `Dst: BikeshedIntrinsicFrom<Src, ASSUMPTIONS>`, the compiler guarantees
+/// that `Src` is soundly *union-transmutable* into a value of type `Dst`,
+/// provided that the programmer has guaranteed that the given
+/// [`ASSUMPTIONS`](Assume) are satisfied.
+///
+/// A union-transmute is any bit-reinterpretation conversion in the form of:
```
I love how concise this is.

One question, though: are the semantics of this operation fully specified? In particular, if `size_of::<Src>() < size_of::<Dst>()`, is it well-defined that the trailing bytes in `Transmute<Src, Dst>` are uninitialized?

If the answer is yes, then no change is required here in response to this comment.

jswrenn replied:
It's how Rust works now, but it's not as well-documented as I'd like — we'll need to cite other official documentation before we stabilize this. That said, I'm confident the decision here is a 'safe' one. `MaybeUninit` relies on these semantics, though theoretically that could change. I think that's unlikely, since the present discussion is whether `union`s should *always* implicitly admit uninitialized bytes, even outside of the padding-extension case.

Should that discussion be rolled back so severely that Rust, say, documents those excess bytes to be initialized, we have a secondary bulwark in @scottmcm's plan to redefine the `transmute_unchecked` to permit padding extension in this same way (it's useful for SIMD).

This is all to say: I think the winds are blowing sufficiently hard in the right direction that we can tentatively adopt these semantics. We'll need to pin them down definitively before stabilization, but that's a long ways off.

joshlf replied:
Sounds reasonable. Can we document this as a blocker somewhere to make sure it's resolved before stabilization?

jswrenn replied:
I've added a note to *Unresolved Questions* in the [tracking issue](https://github.com/rust-lang/rust/issues/99571).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10517058484/job/29140478393) [(plain)](https://github.com/rust-lang/rust/commit/aa01623c2be7f41e75a116eb201d227c10a4f8ff/checks/29140478393/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.60s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- core/src/mem/transmutability.rs - mem::transmutability::Assume::safety (line 229) stdout ----
Test executable failed (exit status: 101).

stderr:
thread 'main' panicked at core/src/mem/transmutability.rs:23:1:
assertion failed: matches!(maybe_dst, Some(EvenU8 { val: 2 }))



failures:
```

</details>

jswrenn commented on library/core/src/mem/transmutability.rs:
````diff
@@ -1,10 +1,80 @@
+/// extend the bits of `Src` with trailing padding to fill trailing
+/// uninitialized bytes of `Self`; e.g.:
+///
+/// ```ignore (cannot doctest this until bootstrap; remove ignore and corresponding test in tests/mem.rs in the future)
````
I went ahead and added more examples, using the trick here: https://github.com/rust-lang/rust/pull/129032#discussion_r1722021422

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 2540070fd45defe48d02aa7af974c8f295b326a4 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 2540070fd45defe48d02aa7af974c8f295b326a4 -->
<!-- homu: {"type":"Approved","sha":"2540070fd45defe48d02aa7af974c8f295b326a4","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
